code/modules/wiremod/components/utility/typecast.dm

```diff
@@ -27,33 +32,86 @@
 		PORT_TYPE_STRING,
 		PORT_TYPE_NUMBER,
 		PORT_COMPOSITE_TYPE_LIST,
+		PORT_COMPOSITE_TYPE_ASSOC_LIST,
 		PORT_TYPE_ATOM,
+		PORT_TYPE_TABLE,
 	)
+
 	typecast_options = add_option_port("Typecast Options", component_options)
 
-/obj/item/circuit_component/typecast/pre_input_received(datum/port/input/port)
-	var/current_option = typecast_options.value
-	if(current_type != current_option)
-		current_type = current_option
-		output_value.set_datatype(current_type)
+	list_type_options = add_option_port("List Type", (GLOB.wiremod_basic_types) - PORT_TYPE_SIGNAL)
 
+/obj/item/circuit_component/typecast/pre_input_received(datum/port/input/port)
+	update_output()
 
 /obj/item/circuit_component/typecast/input_received(datum/port/input/port)
 	var/current_option = typecast_options.value
 	var/value = input_value.value
 	var/value_to_set = null
-	switch(current_option)
-		if(PORT_TYPE_STRING)
-			if(istext(value))
-				value_to_set = value
-		if(PORT_TYPE_NUMBER)
-			if(isnum(value))
-				value_to_set = value
-		if(PORT_COMPOSITE_TYPE_LIST)
-			if(islist(value))
-				value_to_set = value
-		if(PORT_TYPE_ATOM)
-			if(isatom(value))
+
+	if (current_option == PORT_COMPOSITE_TYPE_LIST || current_option == PORT_COMPOSITE_TYPE_ASSOC_LIST)
+		if(islist(value))
+			var/is_correct_format = TRUE
+			var/is_correctly_typed = TRUE
+			var/want_assoc = (current_option == PORT_COMPOSITE_TYPE_ASSOC_LIST)
+
+			for (var/key as anything in value)
+				if (!isnull(value[key]))
+					is_correct_format = want_assoc
+					break
+
+			if (is_correct_format)
+				for (var/entry as anything in value)
+					if (want_assoc)
+						entry = value[entry]
+					if (!entry_valid(entry, list_type_options.value))
+						is_correctly_typed = FALSE
+						break
+
+			if (is_correct_format && is_correctly_typed)
 				value_to_set = value
 
+	else if (entry_valid(value, current_option))
+		value_to_set = value
+
+	update_output()
 	output_value.set_output(value_to_set)
+
+/obj/item/circuit_component/typecast/proc/entry_valid(entry, comparison_type)
+	switch (comparison_type)
+		if (PORT_TYPE_ANY)
+			return TRUE
+		if (PORT_TYPE_ATOM)
+			return isatom(entry)
+		if (PORT_TYPE_NUMBER)
+			return isnum(entry)
+		if (PORT_TYPE_STRING)
+			return istext(entry)
+		if (PORT_TYPE_TABLE)
+			if (islist(entry))
+				if (!length(entry))
+					return FALSE
+				for (var/potential_list as anything in entry)
+					if (!islist(potential_list)) // if any entry isnt a nested list, nope
+						return FALSE
+				return TRUE // only occurs if every entry is a list, and entry itself is a list
+			return FALSE
```

Comment:
No, table types cannot be casted to or from because typecast cannot guarantee the integrity of the data without checking every entry in the sublist too. There may be list list types in the future so this isn't a super reliable check.
In any case, I'd only keep typecasting to the primitive types just for sanity sake, we don't want users casting malformed any types to variables where they won't necessarily fit and cause bugs.